### PR TITLE
Smart pointer notify objects changed to weak pointers (analytics)

### PR DESCRIPTION
**Pull request by rupnikj**
(PNotify -> TWPt<TNotify>) to prevent segmentation faults when node exits

### CODE REVIEW

File: src/glib/mine/svm.h [hunk 1]
```diff
@@ -388,7 +388,7 @@ template <class TVecV>
 TLinModel SolveClassify(const TVecV& VecV, const int& Dims, const int& Vecs,
         const TFltV& TargetV, const double& Cost, const double& UnbalanceWgt,
         const int& MxMSecs, const int& MxIter, const double& MnDiff, 
-        const int& SampleSize, const PNotify& Notify = TStdNotify::New()) {
+        const int& SampleSize, const TWPt<TNotify>& Notify = TNotify::StdNotify()) {
```
Comment (blazf) [inline src/glib/mine/svm.h]: I would remove default option, to be consistant with `SolveRegression` bellow.

File: src/glib/mine/vizmap.cpp [hunk 2]
```diff
@@ -998,7 +998,7 @@ PBowDocBs TVizMapXmlDocBs::LoadBowDocBs(const TStr& XmlFNm, PSwSet SwSet,
 //////////////////////////////////////////////////////////////////////////
 // Visualization-Map Factory
 void TVizMapFactory::CG(const TMatrix& Matrix, const TFltV& b, 
-        TFltV& x, PNotify Notify, const int& MaxStep, const double& EpsTer) {
+        TFltV& x, TWPt<TNotify> Notify, const int& MaxStep, const double& EpsTer) {
```
Comment (blazf) [inline src/glib/mine/vizmap.cpp]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

File: src/glib/mine/vizmap.cpp [hunk 3]
```diff
@@ -1041,7 +1041,7 @@ void TVizMapFactory::CG(const TMatrix& Matrix, const TFltV& b,
 void TVizMapFactory::MakeFlat(const PSVMTrainSet& Set, 
         const TVizDistType& DistType, TVec<TFltV>& DocPointV, 
         const int& MxStep, const int& MxSecs, const double& MnDiff, 
-        const bool& RndStartPos, PNotify& Notify) {
+        const bool& RndStartPos, TWPt<TNotify>& Notify) {
```
Comment (blazf) [inline src/glib/mine/vizmap.cpp]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

File: src/glib/mine/vizmap.cpp [hunk 4]
```diff
@@ -1152,7 +1152,7 @@ void TVizMapFactory::NormalizePoints(TVec<TFltV>& PointV) {
 void TVizMapFactory::LsiMds(TVec<PBowSpV> DocSpV, PSemSpace SemSpace, 
         TVec<TFltV>& DocPointV, const double& SemSpaceThresh, 
         const int& MxStep, const int& MxSecs, const double& MnDiff, 
-        PNotify Notify) {
+        TWPt<TNotify> Notify) {
```
Comment (blazf) [inline src/glib/mine/vizmap.cpp]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

File: src/glib/mine/vizmap.cpp [hunk 5]
```diff
@@ -1177,7 +1177,7 @@ void TVizMapFactory::LsiMds(TVec<PBowSpV> DocSpV, PSemSpace SemSpace,
 PVizMapFrame TVizMapFactory::DocLsiMds(PBowDocWgtBs BowDocWgtBs,
         PSemSpace SemSpace, const TVec<TFltV>& _DocPointV, 
         const double& SemSpaceThresh, const int& MxStep, 
-        const int& MxSecs, const double& MnDiff, PNotify Notify) {
+        const int& MxSecs, const double& MnDiff, TWPt<TNotify> Notify) {
```
Comment (blazf) [inline src/glib/mine/vizmap.cpp]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

File: src/glib/mine/vizmap.cpp [hunk 6]
```diff
@@ -1252,7 +1252,7 @@ void TVizMapFactory::AddDocMetadata(PVizMapFrame VizMapFrame,
 
 void TVizMapFactory::LndMrk(PBowDocWgtBs BowDocWgtBs, PBowSim BowSim,
         const TVec<PBowSpV>& ClustSpV, const TVec<TFltV>& ClustPointV, 
-        TVec<TFltV>& DocPointV, const int& LinCombNum, PNotify Notify) {
+        TVec<TFltV>& DocPointV, const int& LinCombNum, TWPt<TNotify> Notify) {
```
Comment (blazf) [inline src/glib/mine/vizmap.cpp]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

File: src/glib/mine/vizmap.cpp [hunk 7]
```diff
@@ -1290,7 +1290,7 @@ void TVizMapFactory::LndMrk(PBowDocWgtBs BowDocWgtBs, PBowSim BowSim,
 
 PVizMapFrame TVizMapFactory::DocLndMrk(PBowDocWgtBs BowDocWgtBs, 
         const TVec<PBowSpV>& ClustSpV, const TVec<TFltV>& ClustPointV, 
-        const int& LinCombNum, PNotify Notify) {
+        const int& LinCombNum, TWPt<TNotify> Notify) {
```
Comment (blazf) [inline src/glib/mine/vizmap.cpp]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

File: src/glib/mine/vizmap.cpp [hunk 8]
```diff
@@ -1315,7 +1315,7 @@ PVizMapFrame TVizMapFactory::DocLndMrk(PBowDocWgtBs BowDocWgtBs,
 PVizMapFrame TVizMapFactory::ClustLsiMdsDocLndMrk(PBowDocWgtBs BowDocWgtBs,
         PBowDocPart BowDocPart, PSemSpace SemSpace, const int& LinCombNum, 
         const double& SemSpaceThresh, const int& MxStep, const int& MxSecs, 
-        const double& MnDiff, PNotify Notify) {
+        const double& MnDiff, TWPt<TNotify> Notify) {
```
Comment (blazf) [inline src/glib/mine/vizmap.cpp]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

File: src/glib/mine/vizmap.h [hunk 9]
```diff
@@ -525,26 +525,26 @@ typedef enum {vdtEucl, vdtCos, vdtSqrtCos, vdtDistMtx} TVizDistType;
 class TVizMapFactory {
 private:
     static void CG(const TMatrix& Matrix, const TFltV& b, TFltV& x,
-        PNotify Notify, const int& MaxStep = 200, const double& EpsTer = 1e-7);
+        TWPt<TNotify> Notify, const int& MaxStep = 200, const double& EpsTer = 1e-7);
```
Comment (blazf) [inline src/glib/mine/vizmap.h]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

File: src/glib/mine/vizmap.h [hunk 10]
```diff
@@ -525,26 +525,26 @@ typedef enum {vdtEucl, vdtCos, vdtSqrtCos, vdtDistMtx} TVizDistType;
...
 
 public:
     // the ultimate MDS...
     static void MakeFlat(const PSVMTrainSet& Set, const TVizDistType& DistType, 
         TVec<TFltV>& DocPointV, const int& MxStep, const int& MxSecs, 
-        const double& MnDiff, const bool& RndStartPos, PNotify& Notify);
+        const double& MnDiff, const bool& RndStartPos, TWPt<TNotify>& Notify);
```
Comment (blazf) [inline src/glib/mine/vizmap.h]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

File: src/glib/mine/vizmap.h [hunk 11]
```diff
@@ -525,26 +525,26 @@ typedef enum {vdtEucl, vdtCos, vdtSqrtCos, vdtDistMtx} TVizDistType;
...
     static void NormalizePoints(TVec<TFltV>& PointV);
 
     // BowSpV->LSI->MDS->DocPointV
     // applys LSI and MDS to documents in DocSpV
     static void LsiMds(TVec<PBowSpV> DocSpV, PSemSpace SemSpace, TVec<TFltV>& DocPointV, 
         const double& SemSpaceThresh, const int& MxStep, const int& MxSecs, 
-        const double& MnDiff, PNotify Notify);
+        const double& MnDiff, TWPt<TNotify> Notify);
```
Comment (blazf) [inline src/glib/mine/vizmap.h]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

File: src/glib/mine/vizmap.h [hunk 12]
```diff
@@ -525,26 +525,26 @@ typedef enum {vdtEucl, vdtCos, vdtSqrtCos, vdtDistMtx} TVizDistType;
...
 
     // Bow->LSI->MDS->Viz
     // for smaller datasets
     static PVizMapFrame DocLsiMds(PBowDocWgtBs BowDocWgtBs, PSemSpace SemSpace,
         const TVec<TFltV>& _DocPointV, const double& SemSpaceThresh, 
-        const int& MxStep, const int& MxSecs, const double& MnDiff, PNotify Notify);
+        const int& MxStep, const int& MxSecs, const double& MnDiff, TWPt<TNotify> Notify);
```
Comment (blazf) [inline src/glib/mine/vizmap.h]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

File: src/glib/mine/vizmap.h [hunk 13]
```diff
@@ -556,101 +556,101 @@ class TVizMapFactory {
     // on positions ClustPointV of landmarks ClustSpV
     static void LndMrk(PBowDocWgtBs BowDocWgtBs, PBowSim BowSim,
         const TVec<PBowSpV>& ClustSpV, const TVec<TFltV>& ClustPointV, 
-        TVec<TFltV>& DocPointV, const int& LinCombNum, PNotify Notify);
+        TVec<TFltV>& DocPointV, const int& LinCombNum, TWPt<TNotify> Notify);
```
Comment (blazf) [inline src/glib/mine/vizmap.h]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

File: src/glib/mine/vizmap.h [hunk 14]
```diff
@@ -556,101 +556,101 @@ class TVizMapFactory {
...
 
     // Bow->Landmarks->Viz
     static PVizMapFrame DocLndMrk(PBowDocWgtBs BowDocWgtBs, 
         const TVec<PBowSpV>& ClustSpV, const TVec<TFltV>& ClustPointV, 
-        const int& LinCombNum = 3, PNotify Notify = TNullNotify::New());
+        const int& LinCombNum = 3, TWPt<TNotify> Notify = TNullNotify::New());
```
Comment (blazf) [inline src/glib/mine/vizmap.h]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

File: src/glib/mine/vizmap.h [hunk 15]
```diff
@@ -556,101 +556,101 @@ class TVizMapFactory {
...
     // Bow->KMeans->LSI->MDS->Viz -- for larger datasets
     static PVizMapFrame ClustLsiMdsDocLndMrk(PBowDocWgtBs BowDocWgtBs,
         PBowDocPart BowDocPart, PSemSpace SemSpace, const int& LinCombNum, 
         const double& SemSpaceThresh, const int& MxStep, const int& MxSecs, 
-        const double& MnDiff, PNotify Notify);
+        const double& MnDiff, TWPt<TNotify> Notify);
```
Comment (blazf) [inline src/glib/mine/vizmap.h]: `TWPt<TNotify>` => `const TWPt<TNotify>&`

**Comment by rupnikj** [comment]
Fixed

**Pull request merged**